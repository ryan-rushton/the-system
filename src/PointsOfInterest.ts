import { createRef } from 'react';

/**
 * All the points of interest in the solar system that get rendered.
 * Values are obtained from either https://nssdc.gsfc.nasa.gov/planetary/factsheet/ or wikipedia.
 *
 * The following properties may or may not exist on each point of interest,
 * - id: string - A unique string used for data-testid identifiers for tests.
 * - ref: RefObject<HTMLDivElement> - A ref so that we can scroll to the object when it is selected in the nav menu.
 * - display: string - The name of the point.
 * - radius: number - The radius of the point of interest (includes rings if visible), units are km.
 * - distance: number - The points distance from the sun, units are km.
 * - orbitalPeriod: number - How many days it takes the point to orbit the sun.
 * - Satellites: PointOfInterest[] - An array of satellites that orbit the point. These have radius, distance,
 *  orbitalPeriod just like the points.
 */
export const pointsOfInterest = {
  sun: {
    id: 'sun',
    ref: createRef<HTMLDivElement>(),
    radius: 695508,
    distance: 0,
    orbitalPeriod: 1,
  },
  mercury: {
    id: 'mercury',
    ref: createRef<HTMLDivElement>(),
    radius: 2440,
    distance: 57900000,
    orbitalPeriod: 88,
  },
  venus: {
    id: 'venus',
    ref: createRef<HTMLDivElement>(),
    radius: 6052,
    distance: 108200000,
    orbitalPeriod: 224.7,
  },
  earth: {
    id: 'earth',
    ref: createRef<HTMLDivElement>(),
    radius: 6378,
    distance: 149600000,
    orbitalPeriod: 365.2,
    satellites: [
      {
        id: 'moon',
        distance: 384400,
        orbitalPeriod: 27.3,
        radius: 1737,
      },
    ],
  },
  mars: {
    id: 'mars',
    ref: createRef<HTMLDivElement>(),
    radius: 3396,
    distance: 227900000,
    orbitalPeriod: 687,
  },
  theBelt: {
    id: 'theBelt',
    ref: createRef<HTMLDivElement>(),
  },
  jupiter: {
    id: 'jupiter',
    ref: createRef<HTMLDivElement>(),
    radius: 71498,
    distance: 778600000,
    orbitalPeriod: 4331,
    satellites: [
      {
        id: 'io',
        distance: 421700,
        orbitalPeriod: 1.77,
        radius: 1822,
      },
      {
        id: 'europa',
        distance: 670900,
        orbitalPeriod: 3.55,
        radius: 1561,
      },
      {
        id: 'ganymede',
        distance: 1070400,
        orbitalPeriod: 7.15,
        radius: 2634,
      },
      {
        id: 'callisto',
        distance: 1882700,
        orbitalPeriod: 16.69,
        radius: 2410,
      },
    ],
  },
  saturn: {
    id: 'saturn',
    ref: createRef<HTMLDivElement>(),
    radius: 80000 + 58232, // making the ring radius clear
    distance: 1433500000,
    orbitalPeriod: 10747,
    satellites: [
      {
        id: 'titan',
        distance: 1221870,
        orbitalPeriod: 15.95,
        radius: 2574,
      },
      {
        id: 'rhea',
        distance: 527108,
        orbitalPeriod: 1.77,
        radius: 764,
      },
    ],
  },
  uranus: {
    id: 'uranus',
    ref: createRef<HTMLDivElement>(),
    radius: 25559,
    distance: 2872500000,
    orbitalPeriod: 30589,
  },
  neptune: {
    id: 'neptune',
    ref: createRef<HTMLDivElement>(),
    radius: 24764,
    distance: 4495100000,
    orbitalPeriod: 59800,
  },
  pluto: {
    id: 'pluto',
    ref: createRef<HTMLDivElement>(),
    radius: 1185,
    distance: 5906400000,
    orbitalPeriod: 90560,
  },
} as const;
