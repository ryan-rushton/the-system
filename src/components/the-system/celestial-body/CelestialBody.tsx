import clsx from 'clsx';
import { RefObject } from 'react';
import styles from './CelestialBody.module.scss';
import './CelestialBodyAnimations.scss';

export interface CelestialBodyProps {
  /** A value to use on data-testid for cypress tests. Likely the display from pointsOfInterest. */
  id: string;
  /** Class name for the celestial body */
  className?: string;
  /** Distance from the sun with distance multiplier context applied */
  distance: number;
  /** Orbital period for the body orbital period multiplier context applied */
  orbitalPeriod: number;
  /** Radius for the body with size multiplier context applied */
  radius: number;
  /**
   * This is the reference point the left and top calculations are referencing.
   *
   * This should either be the system radius for planets/sun or the radius of whatever the satellite is orbiting.
   * */
  referenceRadius: number;
  /** Whether the red orbit line is visible. */
  hasOrbitLine?: boolean;
  /** Any satellites that are orbiting this body. */
  satellites?: CelestialBodyProps[];
  /** A ref so that we can scroll to this body. */
  scrollToRef?: RefObject<HTMLDivElement>;
}

/**
 * This is the main component that places the elements on the page.
 *
 * It works by using one div for the orbit which is essentially a large disc.
 * On top of this we place the visual element for the body right at the edge.
 * The same edge also gets a border when the orbit lines are visible.
 * Satellite's are also CelestialBody's and are placed underneath the visual element of the body.
 * Basically this whole things is a bunch of spinning discs sitting on top of one another.
 */
export function CelestialBody({
  id,
  className,
  distance,
  orbitalPeriod,
  radius,
  hasOrbitLine,
  referenceRadius,
  satellites,
  scrollToRef,
}: CelestialBodyProps) {
  // orbit calcs and style
  const edgeOffset = referenceRadius - distance - radius;
  const heightWidth = 2 * (distance + radius);

  const orbitStyles = {
    animation: `orbit ${orbitalPeriod}s linear infinite`,
    height: heightWidth,
    left: edgeOffset,
    top: edgeOffset,
    width: heightWidth,
  };

  // body calcs and style
  // Assume something with distance 0 (this sun) is already centred
  const top = distance > 0 ? '50%' : 0;
  // left needs to be negative so the center of the body sits on the orbit line.
  const left = distance > 0 ? -radius : 0;

  const bodyStyle = {
    animation: `planet-rotation ${orbitalPeriod}s linear infinite`,
    height: radius * 2,
    left,
    top,
    width: radius * 2,
  };

  return (
    <div
      data-testid={`${id}-orbit-line`}
      className={clsx(styles.orbit, { orbitLine: hasOrbitLine })}
      style={orbitStyles}
    >
      <div
        data-testid={id}
        className={clsx(styles.celestialBody, className)}
        title={className}
        ref={scrollToRef}
        style={bodyStyle}
      >
        {satellites?.map((satellite) => (
          <CelestialBody
            className={satellite.className}
            distance={satellite.distance}
            hasOrbitLine={hasOrbitLine}
            key={satellite.id}
            orbitalPeriod={satellite.orbitalPeriod}
            radius={satellite.radius}
            referenceRadius={satellite.referenceRadius}
            id={satellite.id}
          />
        ))}
      </div>
    </div>
  );
}
