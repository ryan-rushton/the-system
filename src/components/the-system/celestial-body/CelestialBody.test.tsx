import { render, screen } from '@testing-library/react';
import { describe, expect, test } from 'vitest';
import { pointsOfInterest } from '../../../PointsOfInterest';
import CelestialBody from './CelestialBody';

describe('CelestialBody', () => {
  test('it matches the snapshot', () => {
    const { jupiter } = pointsOfInterest;
    render(
      <CelestialBody
        id={jupiter.id}
        className={'some-styles'}
        distance={jupiter.distance}
        orbitalPeriod={jupiter.orbitalPeriod}
        radius={jupiter.radius}
        referenceRadius={1000}
      />,
    );

    expect(screen.getByTestId('jupiter-orbit-line')).toMatchSnapshot();
  });

  test('it renders the correct number of celestial bodies', () => {
    const { jupiter } = pointsOfInterest;
    render(
      <CelestialBody
        id={jupiter.id}
        className={'some-styles'}
        distance={jupiter.distance}
        orbitalPeriod={jupiter.orbitalPeriod}
        radius={jupiter.radius}
        referenceRadius={1000}
        satellites={jupiter.satellites.map((s) => ({ ...s, referenceRadius: jupiter.radius }))}
      />,
    );

    expect(screen.getByTestId('jupiter')).toBeInTheDocument();
    for (const satellite of jupiter.satellites) {
      expect(screen.getByTestId(satellite.id)).toBeInTheDocument();
    }
  });
});
