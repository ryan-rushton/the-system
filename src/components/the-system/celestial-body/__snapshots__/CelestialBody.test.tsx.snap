// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CelestialBody it matches the snapshot 1`] = `
<div
  class="orbit"
  data-testid="jupiter-orbit-line"
  style="animation: orbit 4331s linear infinite; height: 1557342996px; left: -778670498px; top: -778670498px; width: 1557342996px;"
>
  <div
    class="celestialBody some-styles"
    data-testid="jupiter"
    style="animation: planet-rotation 4331s linear infinite; height: 142996px; left: -71498px; top: 50%; width: 142996px;"
    title="some-styles"
  />
</div>
`;
